index.bs

```diff
@@ -61,16 +61,48 @@ different storage mechanism with a different API for such files. The entry point
 
 A <dfn export id="entry">file system entry</dfn> is either a [=file entry=] or a [=directory entry=].
 
-Each [=/file system entry=] has an associated <dfn for="file system entry" id=entry-query-access>query access</dfn> algorithm, which takes "`read`"
-or "`readwrite`" <var ignore>mode</var> and returns a {{PermissionState}}. Unless specified
-otherwise it returns "{{PermissionState/denied}}". The algorithm is allowed to throw.
-
-Each [=/file system entry=] has an associated <dfn for="file system entry" id=entry-request-access>request access</dfn> algorithm, which takes
-"`read`" or "`readwrite`" <var ignore>mode</var> and returns a {{PermissionState}}. Unless specified
-otherwise it returns "{{PermissionState/denied}}". The algorithm is allowed to throw.
-
-Note: Implementations that only implement this specification and not dependent specifications do not
-need to bother implementing [=/file system entry=]'s [=file system entry/query access=] and [=file system entry/request access=].
+Each [=/file system entry=] has an associated
+<dfn for="file system entry" id=entry-query-access>query access</dfn>
+algorithm, which takes "`read`" or "`readwrite`" <var ignore>mode</var> and
+returns a [=/file system access result=].
+Unless specified otherwise it returns a [=/file system access result=] with a
+[=file system access result/permission state=] of "{{PermissionState/denied}}"
+and with a [=file system access result/error name=] of « the empty string ».
+
+Each [=/file system entry=] has an associated
+<dfn for="file system entry" id=entry-request-access>request access</dfn>
+algorithm, which takes "`read`" or "`readwrite`" <var ignore>mode</var> and
+returns a [=/file system access result=].
+Unless specified otherwise it returns a [=/file system access result=] with a
+[=file system access result/permission state=] of "{{PermissionState/denied}}"
+and with a [=file system access result/error name=] of « the empty string ».
+
+A <dfn export>file system access result</dfn> is a [=struct=] encapsulating the
+result of [=file system entry/query access|querying=] or
+[=file system entry/request access|requesting=] access to the file system.
+It has the following [=struct/items=]:
+
+: <dfn for="file system access result">permission state</dfn>
+:: A {{PermissionState}}
+: <dfn for="file system access result">error name</dfn>
+:: A [=string=] which must be either « the empty string » or an
+   [=exception/error name=] listed in the [=error names table=].
+   If [=file system access result/permission state=] is
+   "{{PermissionState/granted}}" this must be « the empty string »
```

Comment:
Can we make it so that when permission is denied it has to be an error name so we don't have to do defaulting below?

Reply:
Done. This simplifies things quite a lot

Side note: is "iff" okay to use in specs? Happy to spell it out if not